crates/biome_js_analyze/src/lint/nursery/no_restricted_types.rs

```diff
@@ -0,0 +1,356 @@
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct NoRestrictedTypesOptions {
+    extend_defaults: Option<bool>,
+    types: FxHashMap<String, CustomRestrictedType>,
+}
+
+#[derive(Debug, Clone, Default, Deserializable, Deserialize, Serialize, Eq, PartialEq)]
+#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct CustomRestrictedType {
+    message: String,
+    fix_with: Option<String>,
+}
+
+declare_node_union! {
+    pub TsRestrictedType = TsReferenceType | TsObjectType
+}
+
+pub struct State {
+    /// Name of the restricted type.
+    restricted_type: RestrictedType,
+    /// Text range used to diagnostic the banned type.
+    range: TextRange,
+    /// Reference to the node to be replaced in the action.
+    /// This is optional because we don't replace empty objects references.
+    reference_identifier: Option<JsReferenceIdentifier>,
+}
+
+#[derive(Debug)]
+pub enum RestrictedType {
+    BigInt,
+    Boolean,
+    Function,
+    Number,
+    Object,
+    String,
+    Symbol,
+    /// {}
+    EmptyObject,
+    /// User provided custom restricted type
+    Custom((String, CustomRestrictedType)),
+}
+
+impl RestrictedType {
+    /// construct a [RestrictedType] from the textual name of a JavaScript type
+    fn from_str(s: &str) -> Option<Self> {
+        Some(match s {
+            "BigInt" => Self::BigInt,
+            "Boolean" => Self::Boolean,
+            "Function" => Self::Function,
+            "Number" => Self::Number,
+            "Object" => Self::Object,
+            "String" => Self::String,
+            "Symbol" => Self::Symbol,
+            "{}" => Self::EmptyObject,
+            _ => return None,
+        })
+    }
+
+    /// Retrieves a diagnostic message from a [RestrictedType]
+    fn additional_note(&self) -> Option<&str> {
+        Some(match *self {
+			| Self::BigInt
+			| Self::Boolean
+			| Self::Number
+			| Self::String
+			| Self::Symbol => "Use lowercase primitives for consistency.",
+			Self::Function =>
+				"Prefer explicitly define the function shape. This type accepts any function-like value, which can be a common source of bugs.",
+			Self::Object =>
+				"Prefer explicitly define the object shape. This type means \"any non-nullable value\", which is slightly better than 'unknown', but it's still a broad type.",
+			Self::EmptyObject => "Prefer explicitly define the object shape. '{}' means \"any non-nullable value\".",
+            _ => return None,
+        })
+    }
+
+    fn suggested_fix_type(&self) -> Option<String> {
+        Some(match self {
+            Self::BigInt => "bigint".to_string(),
+            Self::Boolean => "boolean".to_string(),
+            Self::Number => "number".to_string(),
+            Self::String => "string".to_string(),
+            Self::Symbol => "symbol".to_string(),
+            Self::Custom((_, data)) => data.fix_with.clone()?,
```

Comment:
We could avoid an unmeet allocation using `std::ops::Cow`.
```suggestion
    fn suggested_fix_type(&self) -> Option<Cow<'static, str>> {
        Some(match self {
            Self::BigInt => Cow::Borrowed("bigint"),
            Self::Boolean => Cow::Borrowed("boolean"),
            Self::Number => Cow::Borrowed("number"),
            Self::String => Cow::Borrowed("string"),
            Self::Symbol => Cow::Borrowed("symbol"),
            Self::Custom((_, data)) => Cow::Owned(data.fix_with.clone()?),
```